### PR TITLE
Lock portmapper per range

### PR DESCRIPTION
**Pull request by LK4D4**
It's longest Lock in docker.

### CODE REVIEW

File: portmapper/mapper.go [hunk 1]
```diff
@@ -122,35 +124,32 @@ func (pm *PortMapper) MapRange(container net.Addr, hostIP net.IP, hostPortStart,
 		}
 	}()
 
+	pm.lock.Lock()
```
Comment (mrjana) [inline portmapper/mapper.go]: Let's not hold the lock around unnecessary code. It just needs to be protect `currentMappings`. in fact this code be rewritten like this:
```
key := getKey(m.host)

pm.lock.Lock()
_, exists := pm.currentMappings[key]
pm.lock.Unlock()

if exists {
     return nil, ErrPortMappedForIP
}
```

**Comment by mrjana** [comment]
Other than one minor comment it LGTM

**Comment by LK4D4** [comment]
@mrjana thanks for comment. Fixed.

**Comment by mrjana** [comment]
LGTM

File: portmapper/mapper.go [hunk 2]
```diff
@@ -67,15 +70,14 @@ func (pm *PortMapper) Map(container net.Addr, hostIP net.IP, hostPort int, usePr
...
-	pm.lock.Lock()
-	defer pm.lock.Unlock()
-
 	var (
 		m                 *mapping
 		proto             string
 		allocatedHostPort int
 	)
-
+	lockKey := fmt.Sprintf("%s:%d-%d", hostIP, hostPortStart, hostPortEnd)
```
Comment (aboch) [inline portmapper/mapper.go]: I am trying to understand what we get by using key lock in this function. You are already breaking the big lock section in small ones where it is needed. 
Do we also need the granular, very specific `hostIP:hostPortStart-hostPortEnd` lock ?

What will happen now  when we get two concurrent calls to `MapRange()` with overlapping host port range (but not same start and/or end host ports), so different `locKey` ex: `127.0.0.1:1000-1024` and `127.0.0.1:999-1023` ?

We will have two independent key locks. But the data manipulated by the two calls is overlapping. Then my next question is what are we trying to protect with the key lock ?

Comment (LK4D4) [inline portmapper/mapper.go]: You're right. It seems like it's impossible to do it correct with the current design :/

**Comment by LK4D4** [comment]
I'm closing it, cause there is no way to avoid global locking without locking each port in range for now.

**Pull request closed without merging**